text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,190 @@
+# Summary
+
+Introduce a full featured, official TiKV (and PD) Rust client. It will be intended to be used as a reference implementation, or to provide C-compatible binding for future clients.
+
+# Motivation
+
+Currently, users of TiKV must use [TiDB's Go Client](https://github.com/pingcap/tidb/blob/master/store/tikv/client.go), which is not well packaged or documented. We would like to ensure that users can easily use TiKV and PD without needing to use TiDB.
+
+We think this would help encourage community participation in the TiKV project and associated libraries. During talks with several potential corporate users we discovered that there was an interest in using TiKV to resolve concerns such as caching and raw key-value stores.
+
+# Detailed design
+## Supported Targets
+
+We will target the `stable` channel of Rust starting in the Rust 2018 edition. We choose to begin with Rust 2018 so we do not need to concern ourselves with an upgrade path.
+
+We will also support the most recent `nightly` version of Rust, but users should not feel the need to reach for stable unless they are already using it.
+
+## Naming
+
+While the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/naming.html) do not perscribe any particular crate name convention.
+
+We choose to name the crate `tikv_client` to conform to the constraints presented to us by the Rust compiler. Cargo permits `tikv-client` and `tikv_client`, but `rustc` does not permit `tikv-client` so we choose to use `tikv_client` to reduce mental overhead.
+
+Choosing to seperate `tikv` and `client` helps potentially unfamiliar users to immediately understand the intent of the package. `tikvclient`, while understandable, is not immediately parsable by a human.
+
+All structures and functions will otherwise follow the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/), some of which will be enforced by `clippy`.
+
+## Installation.
+
+To utilize the client programmatically, users will be able to add the `tikv-client` crate to their `Cargo.toml`'s dependencies. Then they must use the crate with `use tikv_client;`. Unfortunately due to Rust’s naming conventions this inconsistency is in place.
+
+To utilize the command line client, users will be able to install the binary via `cargo install tikv-client`. They will then be able to access the client through the binary `tikv-client`. If they wish for a different name they can alias it in their shell.
+
+## Usage
+
+## Key structs and traits
```

Comment:
I think we don't need to list the interfaces one by one because there are too many and we may still change some. 

Here I suggest just mention Raw mode, Transaction Mode, the associated client, and a simple example of how to read + write a key through Raw or Txn Client.

Reply:
I copied some document from go API as template and modified it a little bit here. Please take another look. 
BTW: After I came up with examples. I have a feeling that the API might need some more changes, especially the transactional API. Will create another commit and discuss it [here](https://github.com/tikv/client-rust/pull/3)